internal/backend/oidc/internal/storage/keys/storage.go

```diff
@@ -84,85 +69,79 @@ func (s *Storage) GetKeySet(context.Context) (*jose.JSONWebKeySet, error) {
 }
 
 // GetPublicKeyByID looks up the public key with the given ID.
-func (s *Storage) GetPublicKeyByID(keyID string) (any, error) {
+func (s *Storage) GetPublicKeyByID(ctx context.Context, keyID string) (any, error) {
 	s.lock.Lock()
 	defer s.lock.Unlock()
 
+	err := s.refreshKey(ctx)
+	if err != nil {
+		return nil, err
+	}
+
 	if s.activeKeySet == nil {
 		return nil, errors.New("no active key set")
 	}
 
-	for _, key := range s.activeKeySet.Keys {
-		if key.KeyID == keyID {
-			return key.Key, nil
-		}
+	idx := slices.IndexFunc(s.activeKeySet, func(key op.Key) bool { return key.ID() == keyID })
+	if idx == -1 {
+		return nil, fmt.Errorf("key not found, ID %q", keyID)
 	}
 
-	return nil, fmt.Errorf("key not found, ID %q", keyID)
+	return s.activeKeySet[idx].Key(), nil
 }
 
 // GetCurrentSigningKey returns the active and currently used signing key.
-func (s *Storage) GetCurrentSigningKey() (*jose.JSONWebKey, error) {
+func (s *Storage) GetCurrentSigningKey(ctx context.Context) (op.SigningKey, error) {
 	s.lock.Lock()
 	defer s.lock.Unlock()
 
+	err := s.refreshKey(ctx)
+	if err != nil {
+		return nil, err
+	}
+
 	if s.currentKey != nil {
 		return s.currentKey, nil
 	}
 
 	return nil, errors.New("no current key")
 }
 
-func (s *Storage) keyRefresher(ctx context.Context, keyCh chan<- jose.SigningKey) error {
-	ticker := s.clock.Ticker(external.KeyRotationInterval)
-	defer ticker.Stop()
+func (s *Storage) refreshKey(ctx context.Context) error {
+	if s.clock.Since(s.lastUpdate) < external.KeyRotationInterval {
+		return nil
+	}
 
-	for {
-		// renew the key
-		privateKey, err := s.generateKey()
-		if err != nil {
-			return fmt.Errorf("failure to generate the key: %w", err)
-		}
+	// renew the key
+	privateKey, err := s.generateKey()
+	if err != nil {
+		return fmt.Errorf("failure to generate the key: %w", err)
+	}
 
-		keyID := uuid.NewString()
+	keyID := uuid.NewString()
 
-		if err = s.storeKey(ctx, keyID, privateKey); err != nil {
-			return fmt.Errorf("failure to store the key: %w", err)
-		}
+	ctx = actor.MarkContextAsInternalActor(ctx)
```

Comment:
@Unix4ever @utkuozdemir I'm not sure about this part, but we never pass this context anywhere else. Without this the state below complains about "no access is permitted".

Reply:
It is required for `storeKey` to work.

In the old (current) code, if you trace the ctx passed to `storeKey` all the way to where it originated from (yo go into the library and back out), you'll see that is it the main context marked as the internal actor already.

Probably something changed in the library where they don't propagate it all the way here anymore.

Reply:
The OIDC library no longer starts its own gorotuines. Which means I had to redo logic a bit - now public and sign requests will check for expiration instead of relying on other gorotuine to set the required field in the Storage struct.

Reply:
@utkuozdemir @Unix4ever restored old logic where key rotation is independent of key requests.